File: backend-springboot/src/main/java/com/aura_card/backend/service/SavedCardService.java
package com.aura_card.backend.service;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;

import com.aura_card.backend.dto.request.CreateSavedCardRequest;
import com.aura_card.backend.dto.request.UpdateSavedCardRequest;
import com.aura_card.backend.dto.response.SavedCardDetailResponse;
import com.aura_card.backend.dto.response.SavedCardListItemResponse;
import com.aura_card.backend.exception.AppException;
import com.aura_card.backend.exception.ErrorCode;
import com.aura_card.backend.mapper.SavedCardMapper;
import com.aura_card.backend.model.Card;
import com.aura_card.backend.model.SavedCard;
import com.aura_card.backend.repository.CardRepository;
import com.aura_card.backend.repository.SavedCardRepository;

@Service
public class SavedCardService {
    private final SavedCardRepository savedCardRepository;
    private final CardRepository cardRepository;
    private final SavedCardMapper savedCardMapper;

    public SavedCardService(SavedCardRepository savedCardRepository,
            CardRepository cardRepository,
            SavedCardMapper savedCardMapper) {
        this.savedCardRepository = savedCardRepository;
        this.cardRepository = cardRepository;
        this.savedCardMapper = savedCardMapper;
    }

    // Thêm thiệp lưu
    public SavedCardDetailResponse createSavedCard(CreateSavedCardRequest request, String userId) {
        Card card = cardRepository.findById(request.getCardId())
                .orElseThrow(() -> new AppException(ErrorCode.CARD_NOT_FOUND));

        SavedCard savedCard = savedCardMapper.toSavedCard(request, userId);
        return savedCardMapper.toDetailResponse(savedCardRepository.save(savedCard), card);
    }

    // Cập nhật thiệp lưu
    public SavedCardDetailResponse updateSavedCard(String savedCardId, UpdateSavedCardRequest request) {
        SavedCard savedCard = savedCardRepository.findById(savedCardId)
                .orElseThrow(() -> new AppException(ErrorCode.SAVED_CARD_NOT_FOUND));

        Card card = cardRepository.findById(request.getCardId())
                .orElseThrow(() -> new AppException(ErrorCode.CARD_NOT_FOUND));

        savedCardMapper.updateSavedCard(savedCard, request);
        return savedCardMapper.toDetailResponse(savedCardRepository.save(savedCard), card);
    }

    // Lấy thiệp lưu theo id
    public SavedCardDetailResponse getSavedCardById(String savedCardId) {
        SavedCard savedCard = savedCardRepository.findById(savedCardId)
                .orElseThrow(() -> new AppException(ErrorCode.SAVED_CARD_NOT_FOUND));

        Card card = cardRepository.findById(savedCard.getCardId())
                .orElseThrow(() -> new AppException(ErrorCode.CARD_NOT_FOUND));

        return savedCardMapper.toDetailResponse(savedCard, card);
    }

    // Lấy danh sách thiệp lưu của user (phân trang, limit 12)
    public Page<SavedCardListItemResponse> getSavedCards(String userId, int page, int limit) {
        Pageable pageable = PageRequest.of(page - 1, limit, Sort.by("createdAt").descending());

        return savedCardRepository.findByUserId(userId, pageable)
                .map(savedCard -> {
                    Card card = cardRepository.findById(savedCard.getCardId())
                            .orElseThrow(() -> new AppException(ErrorCode.CARD_NOT_FOUND));
                    return savedCardMapper.toListItemResponse(savedCard, card);
                });
    }

    // Xóa thiệp lưu theo id
    public void deleteSavedCard(String savedCardId) {
        if (!savedCardRepository.existsById(savedCardId)) {
            throw new AppException(ErrorCode.SAVED_CARD_NOT_FOUND);
        }
        savedCardRepository.deleteById(savedCardId);
    }

}

File: backend-nodejs/services/savedcard.service.js
const SavedCard = require("../models/savedCard.model");
const Card = require("../models/card.model");
const AppError = require("../utils/app.error");
const ErrorCode = require("../utils/error.code");

// Thêm thiệp lưu
const createSavedCard = async (
  {
    customName,
    customContent,
    fontFamily,
    fontWeight,
    fontStyle,
    fontColor,
    cardId,
  },
  userId,
) => {
  // kiểm tra card có tồn tại không
  const card = await Card.findById(cardId);
  if (!card) throw new AppError(ErrorCode.CARD_NOT_FOUND);

  const savedCard = await SavedCard.create({
    customName,
    customContent,
    fontFamily,
    fontWeight,
    fontStyle,
    fontColor,
    cardId,
    userId,
  });

  return formatDetailResponse(savedCard, card);
};

// Cập nhật thiệp lưu
const updateSavedCard = async (
  savedCardId,
  {
    customName,
    customContent,
    fontFamily,
    fontWeight,
    fontStyle,
    fontColor,
    cardId,
  },
) => {
  const savedCard = await SavedCard.findById(savedCardId);
  if (!savedCard) throw new AppError(ErrorCode.SAVED_CARD_NOT_FOUND);

  const card = await Card.findById(cardId);
  if (!card) throw new AppError(ErrorCode.CARD_NOT_FOUND);

  savedCard.customName = customName;
  savedCard.customContent = customContent;
  savedCard.fontFamily = fontFamily;
  savedCard.fontWeight = fontWeight;
  savedCard.fontStyle = fontStyle;
  savedCard.fontColor = fontColor;
  savedCard.cardId = cardId;

  await savedCard.save();
  return formatDetailResponse(savedCard, card);
};

// Lấy thiệp lưu theo id
const getSavedCardById = async (savedCardId) => {
  const savedCard = await SavedCard.findById(savedCardId);
  if (!savedCard) throw new AppError(ErrorCode.SAVED_CARD_NOT_FOUND);

  const card = await Card.findById(savedCard.cardId);
  if (!card) throw new AppError(ErrorCode.CARD_NOT_FOUND);

  return formatDetailResponse(savedCard, card);
};

// Lấy danh sách thiệp lưu của user
const getSavedCards = async (userId, { page = 1, limit = 6 }) => {
  const skip = (page - 1) * limit;

  const [savedCards, total] = await Promise.all([
    SavedCard.find({ userId })
      .sort({ createdAt: -1 })
      .skip(skip)
      .limit(Number(limit)),
    SavedCard.countDocuments({ userId }),
  ]);

  // lấy card cho từng savedCard
  const data = await Promise.all(
    savedCards.map(async (savedCard) => {
      const card = await Card.findById(savedCard.cardId);
      if (!card) throw new AppError(ErrorCode.CARD_NOT_FOUND);
      return formatListItemResponse(savedCard, card);
    }),
  );

  return { data, total, totalPages: Math.ceil(total / limit) };
};

// Xóa thiệp lưu
const deleteSavedCard = async (savedCardId) => {
  const savedCard = await SavedCard.findById(savedCardId);
  if (!savedCard) throw new AppError(ErrorCode.SAVED_CARD_NOT_FOUND);
  await savedCard.deleteOne();
};

const formatDetailResponse = (savedCard, card) => ({
  savedCardId: savedCard._id,
  customName: savedCard.customName,
  customContent: savedCard.customContent,
  fontFamily: savedCard.fontFamily,
  fontWeight: savedCard.fontWeight,
  fontStyle: savedCard.fontStyle,
  fontColor: savedCard.fontColor,
  createdAt: savedCard.createdAt,
  updatedAt: savedCard.updatedAt,
  card: {
    cardId: card._id,
    frontImage: card.frontImage,
    backImage: card.backImage,
  },
});

const formatListItemResponse = (savedCard, card) => ({
  savedCardId: savedCard._id,
  customName: savedCard.customName,
  createdAt: savedCard.createdAt,
  card: {
    cardId: card._id,
    frontImage: card.frontImage,
    backImage: card.backImage,
  },
});

module.exports = {
  createSavedCard,
  updateSavedCard,
  getSavedCardById,
  getSavedCards,
  deleteSavedCard,
};
